# Attempt to run NVIDIA SSD impl...

## Obsah

*   [Predspracovanie](#predspracovanie)
<a href='#predspracovanie'> </a>
*   [Statistiky na staticku kvant.](#stats)
<a href='#stats'> </a>
*   [Base precision](#base)
<a href='#base'> </a>
*   [8 bit kvantizacia](#8bit)
<a href='#8bit'> </a>
*   [7 bit kvantizacia](#7bit)
<a href='#8bit'> </a>
*   [6 bit kvantizacia](#6bit)
<a href='#8bit'> </a>
*   [5 bit kvantizacia](#5bit)
<a href='#8bit'> </a>
*   [4 bit kvantizacia](#4bit)
<a href='#4bit'> </a>

<a id='predspracovanie'> </a>
# Predspracovanie

In [1]:
import torch 
precision = 'fp32'
ssd_model = torch.hub.load('NVIDIA/DeepLearningExamples:torchhub', 'nvidia_ssd', model_math=precision)

Using cache found in /home/bohumil/.cache/torch/hub/NVIDIA_DeepLearningExamples_torchhub


<a id='base'> </a>

## Custom base prec. - distiller resnet

In [9]:
! python ./main.py --backbone resnet50 --mode evaluation --checkpoint ./nvidia_ssdpyt_fp32_20190225.pt --data /home/bohumil/FIIT/BP/BP/Zdroje_kod/coco --eval-batch-size 10 --num-workers 1 --split 0.5

Using seed = 4095
loading annotations into memory...
Done (t=0.63s)
creating index...
index created!
loading model checkpoint ./nvidia_ssdpyt_fp32_20190225.pt
Logging configured successfully
Parsing batch: 247/248
Predicting Ended, total time: 485.93 s
Loading and preparing results...
Converting ndarray to lists...
(162609, 7)
0/162609
DONE (t=1.23s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=26.47s).
Accumulating evaluation results...
DONE (t=3.23s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.131
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.218
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.133
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.046
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.140
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.212
 A

## Base prec - original resnet

In [2]:
! python ./main.py --backbone resnet50 --mode evaluation --checkpoint ./nvidia_ssdpyt_fp32_20190225.pt --data /home/bohumil/FIIT/BP/BP/Zdroje_kod/coco --eval-batch-size 10 --num-workers 1 --split 0.5

Using seed = 4298
loading annotations into memory...
Done (t=0.57s)
creating index...
index created!
/home/bohumil/FIIT/BP/BP/Zdroje_kod/dist_env/lib/python3.7/site-packages/torch/nn/_reduction.py:43: UserWarning: size_average and reduce args will be deprecated, please use reduction='none' instead.
  warnings.warn(warning.format(ret))
loading model checkpoint ./nvidia_ssdpyt_fp32_20190225.pt
Parsing batch: 247/248
Predicting Ended, total time: 546.14 s
Loading and preparing results...
Converting ndarray to lists...
(162584, 7)
0/162584
DONE (t=1.47s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=31.65s).
Accumulating evaluation results...
DONE (t=4.52s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.127
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.213
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.130
 Average Precision  (AP) @[ IoU=0.50:0

output:

```python
Using seed = 4298
loading annotations into memory...
Done (t=0.57s)
creating index...
index created!
/home/bohumil/FIIT/BP/BP/Zdroje_kod/dist_env/lib/python3.7/site-packages/torch/nn/_reduction.py:43: UserWarning: size_average and reduce args will be deprecated, please use reduction='none' instead.
  warnings.warn(warning.format(ret))
loading model checkpoint ./nvidia_ssdpyt_fp32_20190225.pt
Parsing batch: 247/248
Predicting Ended, total time: 546.14 s
Loading and preparing results...
Converting ndarray to lists...
(162584, 7)
0/162584
DONE (t=1.47s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=31.65s).
Accumulating evaluation results...
DONE (t=4.52s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.127
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.213
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.130
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.039
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.139
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.216
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.117
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.171
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.180
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.059
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.194
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.289
Current AP: 0.12708
Model precision 0.12708331381314636 mAP
```

<a id='stats'> </a>

## get stats for model with distiller resnet50

In [10]:
! python ./main.py --backbone resnet50 --mode stats --checkpoint ./nvidia_ssdpyt_fp32_20190225.pt --data /home/bohumil/FIIT/BP/BP/Zdroje_kod/coco --eval-batch-size 10 --num-workers 1 --split 0.1

Using seed = 5682
loading annotations into memory...
Done (t=0.50s)
creating index...
index created!
loading model checkpoint ./nvidia_ssdpyt_fp32_20190225.pt
Logging configured successfully
Logging configured successfully
Generating quantization calibration stats based on 0.1 users
Parsing batch: 49/50
Predicting Ended, total time: 926.50 s
Loading and preparing results...
Converting ndarray to lists...
(31898, 7)
0/31898
DONE (t=0.09s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=10.33s).
Accumulating evaluation results...
DONE (t=1.53s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.029
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.045
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.031
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.007
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] =

<a id='8bit'> </a>

# Quantization

## 8 bit quant

In [4]:
! python ./main.py --backbone resnet50 --mode evaluation --quantize --config ./ssd_coco_post_train.yaml  --checkpoint ./nvidia_ssdpyt_fp32_20190225.pt --data /home/bohumil/FIIT/BP/BP/Zdroje_kod/coco --eval-batch-size 5 --num-workers 1 --split 0.5 --no-cuda

Using seed = 6827
loading annotations into memory...
Done (t=0.55s)
creating index...
index created!
loading model checkpoint ./nvidia_ssdpyt_fp32_20190225.pt
Logging configured successfully
Reading configuration from: ./ssd_coco_post_train.yaml
Found component of class PostTrainLinearQuantizer: Name: post_train_quantizer ; Section: quantizers
Loading activation stats from: ./act_quantization_stats.yaml
Preparing model for quantization using PostTrainLinearQuantizer
^C
Traceback (most recent call last):
  File "./main.py", line 328, in <module>
    train(logger, args)
  File "./main.py", line 255, in train
    quantizer.prepare_model(dummy)
  File "/home/bohumil/FIIT/BP/BP/Zdroje_kod/distiller/distiller/quantization/range_linear.py", line 1969, in prepare_model
    super(PostTrainLinearQuantizer, self).prepare_model(dummy_input)
  File "/home/bohumil/FIIT/BP/BP/Zdroje_kod/distiller/distiller/quantization/quantizer.py", line 233, in prepare_model
    summary_graph = distiller.SummaryGra

```
Using seed = 3474
loading annotations into memory...
Done (t=0.51s)
creating index...
index created!
loading model checkpoint ./nvidia_ssdpyt_fp32_20190225.pt
Logging configured successfully
Reading configuration from: ./ssd_coco_post_train.yaml
Found component of class PostTrainLinearQuantizer: Name: post_train_quantizer ; Section: quantizers
Loading activation stats from: /home/bohumil/FIIT/BP/BP/Zdroje_kod/quantization_jupyters/DeepLearningExamples/PyTorch/Detection/SSD/act_quantization_stats.yaml
Preparing model for quantization using PostTrainLinearQuantizer
Applying batch-norm folding ahead of post-training quantization
Propagating output statistics from BN modules to folded modules
Optimizing output statistics for modules followed by ReLU/Tanh/Sigmoid
Updated stats saved to ./quant_stats_after_prepare_model.yaml
Per-layer quantization parameters saved to ./layer_quant_params.yaml
Parsing batch: 495/496
Predicting Ended, total time: 2779.94 s
Loading and preparing results...
Converting ndarray to lists...
(174579, 7)
0/174579
DONE (t=1.24s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=25.23s).
Accumulating evaluation results...
DONE (t=3.18s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.123
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.205
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.127
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.033
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.133
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.202
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.116
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.169
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.179
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.057
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.197
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.271
Current AP: 0.12273
Model precision 0.12273303465396057 mAP
```

<a id='7bit'> </a>

## 7 bit quant

In [2]:
! python ./main.py --backbone resnet50 --mode evaluation --quantize --config ./ssd_coco_post_train7bit.yaml  --checkpoint ./nvidia_ssdpyt_fp32_20190225.pt --data /home/bohumil/FIIT/BP/BP/Zdroje_kod/coco --eval-batch-size 5 --num-workers 1 --split 0.5 --no-cuda --seed 4298

^C
Traceback (most recent call last):
  File "/home/bohumil/FIIT/BP/BP/Zdroje_kod/dist_env/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3023, in _dep_map
    return self.__dep_map
  File "/home/bohumil/FIIT/BP/BP/Zdroje_kod/dist_env/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2817, in __getattr__
    raise AttributeError(attr)
AttributeError: _DistInfoDistribution__dep_map

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./main.py", line 22, in <module>
    from distiller.data_loggers import collector_context
  File "/home/bohumil/FIIT/BP/BP/Zdroje_kod/distiller/distiller/__init__.py", line 39, in <module>
    __version__ = pkg_resources.require("distiller")[0].version
  File "/home/bohumil/FIIT/BP/BP/Zdroje_kod/dist_env/lib/python3.7/site-packages/pkg_resources/__init__.py", line 901, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/home/bohumil/FIIT/BP/BP/Zd

<a id='6bit'> </a>

## 6 bit quant

In [2]:
! python ./main.py --backbone resnet50 --mode evaluation --quantize --config ./ssd_coco_post_train6bit.yaml  --checkpoint ./nvidia_ssdpyt_fp32_20190225.pt --data /home/bohumil/FIIT/BP/BP/Zdroje_kod/coco --eval-batch-size 5 --num-workers 1 --split 0.5 --no-cuda --seed 4298

Using seed = 4298
loading annotations into memory...
Done (t=0.52s)
creating index...
index created!
loading model checkpoint ./nvidia_ssdpyt_fp32_20190225.pt
Logging configured successfully
Reading configuration from: ./ssd_coco_post_train6bit.yaml
Found component of class PostTrainLinearQuantizer: Name: post_train_quantizer ; Section: quantizers
Loading activation stats from: /home/bohumil/FIIT/BP/BP/Zdroje_kod/quantization_jupyters/DeepLearningExamples/PyTorch/Detection/SSD/act_quantization_stats.yaml
Preparing model for quantization using PostTrainLinearQuantizer
Applying batch-norm folding ahead of post-training quantization
Propagating output statistics from BN modules to folded modules
Optimizing output statistics for modules followed by ReLU/Tanh/Sigmoid
Updated stats saved to ./quant_stats_after_prepare_model.yaml
Per-layer quantization parameters saved to ./layer_quant_params.yaml
Parsing batch: 495/496
Predicting Ended, total time: 2381.93 s
Loading and preparing results...


<a id='5bit'> </a>

## 5 bit quant

In [3]:
! python ./main.py --backbone resnet50 --mode evaluation --quantize --config ./ssd_coco_post_train5bit.yaml  --checkpoint ./nvidia_ssdpyt_fp32_20190225.pt --data /home/bohumil/FIIT/BP/BP/Zdroje_kod/coco --eval-batch-size 5 --num-workers 1 --split 0.5 --no-cuda --seed 4298

Using seed = 4298
loading annotations into memory...
Done (t=0.58s)
creating index...
index created!
loading model checkpoint ./nvidia_ssdpyt_fp32_20190225.pt
Logging configured successfully
Reading configuration from: ./ssd_coco_post_train5bit.yaml
Found component of class PostTrainLinearQuantizer: Name: post_train_quantizer ; Section: quantizers
Loading activation stats from: /home/bohumil/FIIT/BP/BP/Zdroje_kod/quantization_jupyters/DeepLearningExamples/PyTorch/Detection/SSD/act_quantization_stats.yaml
Preparing model for quantization using PostTrainLinearQuantizer
Applying batch-norm folding ahead of post-training quantization
Propagating output statistics from BN modules to folded modules
Optimizing output statistics for modules followed by ReLU/Tanh/Sigmoid
Updated stats saved to ./quant_stats_after_prepare_model.yaml
Per-layer quantization parameters saved to ./layer_quant_params.yaml
Parsing batch: 495/496
Predicting Ended, total time: 2388.87 s
Loading and preparing results...


<a id='4bit'> </a>

## 4 bit quant
*   len 0.1 z datasetu

In [2]:
! python ./main.py --backbone resnet50 --mode evaluation --quantize --config ./ssd_coco_post_train_4bit.yaml  --checkpoint ./nvidia_ssdpyt_fp32_20190225.pt --data /home/bohumil/FIIT/BP/BP/Zdroje_kod/coco --eval-batch-size 5 --num-workers 1 --split 0.1 --no-cuda

Using seed = 6671
loading annotations into memory...
Done (t=0.67s)
creating index...
index created!
loading model checkpoint ./nvidia_ssdpyt_fp32_20190225.pt
Logging configured successfully
Reading configuration from: ./ssd_coco_post_train_4bit.yaml
Found component of class PostTrainLinearQuantizer: Name: post_train_quantizer ; Section: quantizers
Loading activation stats from: /home/bohumil/FIIT/BP/BP/Zdroje_kod/quantization_jupyters/DeepLearningExamples/PyTorch/Detection/SSD/act_quantization_stats.yaml
Preparing model for quantization using PostTrainLinearQuantizer
Applying batch-norm folding ahead of post-training quantization
Propagating output statistics from BN modules to folded modules
Optimizing output statistics for modules followed by ReLU/Tanh/Sigmoid
Updated stats saved to ./quant_stats_after_prepare_model.yaml
Per-layer quantization parameters saved to ./layer_quant_params.yaml
Parsing batch: 98/99
Predicting Ended, total time: 555.69 s
Loading and preparing results...
Co

## 4 bit vacsi dataset
*   skúsme 0.5 z datasetu
*   base prec máme na 0.5
*   taktiez nastavime rovnaky seed - 4298

In [2]:
! python ./main.py --backbone resnet50 --mode evaluation --quantize --config ./ssd_coco_post_train_4bit.yaml  --checkpoint ./nvidia_ssdpyt_fp32_20190225.pt --data /home/bohumil/FIIT/BP/BP/Zdroje_kod/coco --eval-batch-size 5 --num-workers 1 --split 0.5 --no-cuda --seed 4298

Using seed = 4298
loading annotations into memory...
Done (t=0.71s)
creating index...
index created!
loading model checkpoint ./nvidia_ssdpyt_fp32_20190225.pt
Logging configured successfully
Reading configuration from: ./ssd_coco_post_train_4bit.yaml
Found component of class PostTrainLinearQuantizer: Name: post_train_quantizer ; Section: quantizers
Loading activation stats from: /home/bohumil/FIIT/BP/BP/Zdroje_kod/quantization_jupyters/DeepLearningExamples/PyTorch/Detection/SSD/act_quantization_stats.yaml
Preparing model for quantization using PostTrainLinearQuantizer
Applying batch-norm folding ahead of post-training quantization
Propagating output statistics from BN modules to folded modules
Optimizing output statistics for modules followed by ReLU/Tanh/Sigmoid
Updated stats saved to ./quant_stats_after_prepare_model.yaml
Per-layer quantization parameters saved to ./layer_quant_params.yaml
Parsing batch: 495/496
Predicting Ended, total time: 2804.28 s
Loading and preparing results...

## 4 bit summary

čiže máme AP = 4%

pre porovnanie, pri full prec. AP = 13%

Teda máme pokles presnosti o cca 66%
 